lib/src/git.rs

```diff
@@ -1319,16 +1322,45 @@ fn push_refs(
             GitPushError::InternalGitError(err)
         }
     })?;
-    let mut remaining_remote_refs: HashSet<_> = qualified_remote_refs.iter().copied().collect();
+    let mut remaining_remote_refs_expected_locations =
+        qualified_remote_refs_expected_locations.clone();
     let mut push_options = git2::PushOptions::new();
     let mut proxy_options = git2::ProxyOptions::new();
     proxy_options.auto();
     push_options.proxy_options(proxy_options);
     let mut callbacks = callbacks.into_git();
+    callbacks.push_negotiation(|updates| {
+        for update in updates {
+            let dst_refname = update
+                .dst_refname()
+                .expect("Expect reference name to be valid UTF-8");
+            let expected_location = (*qualified_remote_refs_expected_locations
+                .get(dst_refname)
+                .expect("Push is trying to move a ref it wasn't asked to move"))
```

Comment:
nit: suppose these data come from remote, it's probably better to return error instead of panicking.

Reply:
Never mind. It appears that the `updates` list is created from `push->specs` (which I think is local data.)

Reply:
I am less clear about the non-utf8 case, but I'm assuming some other part of jj would have panicked earlier if there was a non-utf8 ref floating around.

Reply:
The refspec is constructed by jj, so it should be valid UTF-8 I suppose.